### ISSUE TITLE
Add ignoreOverridden option for ImplicitUnitReturnType

### ISSUE THREAD

**Issue by vladimirfx** [detekt/detekt#2906]
## Expected Behavior of the rule
Such code can't implicitly change return type because it is fixed by interface

    interface DeviceRepository {
        fun createDeviceEntry(deviceId: String, type: DeviceActivity.Type)
    }

    class DeviceRepositoryImpl  {
        override fun createDeviceEntry(
            deviceId: String,
            type: DeviceActivity.Type
        ) = somePrivateImpl(deviceId, type, null)
        ....
    }

**Comment by cortinico** [comment]
Thanks for the report @vladimirfx 

I think this makes sense to be added to the rule (i.e. `override` functions should be excluded from the inspection).

**Comment by schalkms** [comment]
I think this should be the default behavior. There’s no config needed in my opinion. When do you want to not flag the mentioned code?

**Comment by vladimirfx** [comment]
> I think this should be the default behavior. There’s no config needed in my opinion. When do you want to not flag the mentioned code?

Actually, I can’t imagine case where it should be flagged. Rule author may known best...

**Comment by schalkms** [comment]
I meant the case for overridden function. The user could use the curly braces {} syntax as a workaround.

**Comment by arturbosch** [comment]
If the interface return type changes, this code's return type changes too. This can be problematic. 
As a library author want to be aware of this, don't I?

**Comment by vladimirfx** [comment]
> If the interface return type changes, this code's return type changes too. This can be problematic if I use `DeviceRepositoryImpl` directly but not `DeviceRepository` at all or if it is internal and I have to depend only on the implementation.
> As a library author want to be aware of this, don't I?

If interface return type changes then implementation become not compilable anymore. Is it not enough to pay attention? 
In my opinion the whole rule is about IMPLICIT return type change. That not possible for contracted (with interface) methods.

Anyway, you know best, option is enough for me.

**Comment by cortinico** [comment]
> If the interface return type changes, this code's return type changes too. This can be problematic if I use DeviceRepositoryImpl directly but not DeviceRepository at all or if it is internal and I have to depend only on the implementation.

> If interface return type changes then implementation become not compilable anymore. Is it not enough to pay attention?

@arturbosch is right here. For instance in a code similar to your example @vladimirfx:

```kotlin
interface DeviceRepository {
    fun createDeviceEntry(deviceId: String)
}
class DeviceRepositoryImpl : DeviceRepository {
    override fun createDeviceEntry(deviceId: String) = println()
}
```

The return type of both `createDeviceEntry` function is `Unit`. But if you change the interface function to return `Any`, the implementation is still compatible. 

Here having an explicit block will probably be beneficial to clarify the type restriction. However it's still sort of edge case (super return type extended from `Unit` to `Any`) and it's probably the only one for this specific addition.

**Comment by vladimirfx** [comment]
So option is optimal solution.

**Comment by schalkms** [comment]
Uh, right. Artur mentioned the case, where we want the user to notify regarding the change.
After reading the docs and thinking about the rules context, I have to revert my statement from before.

> I think this should be the default behavior. There’s no config needed in my opinion. When do you want to not flag the mentioned code?

The docs mention the following.

> Changing the type of the expression accidentally, changes the functions return type.
> This may lead to backward incompatibility.
> Use a block statement to make clear this function will never return a value.

### Proposal

The rule stays as it is due to the arguments made. Maybe we should state the example in the rule description.
If the user doesn't want to flag the code, there are the following options sorted by usability.

1. Use a block statement
2. Ignore annotation
3. Create a custom rule
4. Disable the rule

-----

> So option is optimal solution.

I don't think so, because it undermines the reasons for writing the rule.
Please also keep in mind the added complexity.

**Comment by vladimirfx** [comment]
IMO the case is very rare. In library if implementation used directly + return type changed in interface from Unit to Any. Anyone has such case in practice?
But when we enabled this rule in one of our small project we have 5 issues in implementation classes.
Disable rule at all? But it is really useful in places where return type is not restricted by interface. 

Maybe it is a bad practice to use expression body syntax for Unit functions?

**Comment by schalkms** [comment]
Thanks for sharing your real world project experience.

> Maybe it is a bad practice to use expression body syntax for Unit functions?

I'd argue that. It's not obvious on the first glance, which type is actually returned.
Anyway, the expression body syntactic sugar is useful only in a few scenarios.

**Comment by arturbosch** [comment]
> IMO the case is very rare. In library if implementation used directly + return type changed in interface from Unit to Any. Anyone has such case in practice?
> But when we enabled this rule in one of our small project we have 5 issues in implementation classes.
> Disable rule at all? But it is really useful in places where return type is not restricted by interface.
> 
> Maybe it is a bad practice to use expression body syntax for Unit functions?

Yep, that is the idea of this rule. Imagine a review situation where you do not have type hints and IDE like features.
Using a block makes the code for Unit return types clearer.

I think I've got the idea for this rule from an episode of the talking kotlin podcast - https://talkingkotlin.com/womens-day-special-idiomatic-kotlin/

**Comment by schalkms** [comment]
Closing this now due to the arguments made for keeping the rule semantics.
